### PR TITLE
New interface for batch preconditioners and batch scaling

### PR DESCRIPTION
**Pull request by Slaedr**
A new interface is introduced for batch preconditioning and scaling. This is done for more consistency with the rest of Ginkgo and to allow more complex preconditioners.

### CODE REVIEW

**Comment by Slaedr** [comment]
format!

**Comment by Slaedr** [comment]
format!

**Review by pratikvn — APPROVED** [review]
Really excellent work! The unified batch solver tests look really good! Overall mostly looks good except a few minor things:
1. Does the trsv need to have some parameters settable from outside ? 
2. For the stopping criteria, it might make sense to combine them into a single struct and pass them in instead of separate parameters duplicated for each solver.

File: common/cuda_hip/preconditioner/batch_identity.hpp.inc [hunk 1]
```diff
@@ -112,7 +115,7 @@ __global__ void batch_identity(
 
         extern __shared__ char local_mem[];
         const auto prec_work = reinterpret_cast<ValueType*>(local_mem);
-        prec.generate(a_b, prec_work);
+        prec.generate(batch, a_b, prec_work);
```
[the remark this answers is not in the document]
Comment (Slaedr) [reply common/cuda_hip/preconditioner/batch_identity.hpp.inc:118]: Yeah, here it's not needed. But In general, that's how the solver is going to call it, so this makes sense for testing the preconditioner.

File: cuda/preconditioner/batch_ilu_kernels.cu [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                 matrix::BatchDense<ValueType>* const z)
+{
+    const auto num_rows = static_cast<int>(l->get_size().at(0)[0]);
+    const auto nbatch = l->get_num_batch_entries();
+    const auto l_ub = get_batch_struct(l);
+    const auto u_ub = get_batch_struct(u);
+    using d_value_type = cuda_type<ValueType>;
+    using trsv_type = BatchExactTrsvSeparate<d_value_type>;
+    using prec_type = BatchIluSplit<d_value_type, trsv_type>;
+    prec_type prec(l_ub, u_ub, trsv_type());
```
[the remark this answers is not in the document]
Comment (Slaedr) [reply cuda/preconditioner/batch_ilu_kernels.cu:100]: If that is the case, the dispatcher should be able to handle that for use in solvers. I guess we will have more branches like [this one](https://github.com/ginkgo-project/ginkgo/blob/130286e4e0edfee308f53964a89a2a82387fb86f/core/solver/batch_dispatch.hpp#L214). This function is just for testing the preconditioners (I doubt anyone would be want to apply them batch-wide in a function call); we can change the kernel call if we need to.

**Pull request merged**